In [1]:
import pandas as pd

titanic = pd.read_csv("titanic.csv")

**문제 1: 열 선택**
다음 작업을 수행하는 코드를 작성하세요.

1. 'PassengerId', 'Name', 'Survived' 세 개 컬럼만 선택하여 첫 10행을 출력하세요.

여러 컬럼을 선택하기 위해 컬럼 이름을 한번 더 대괄호로 감싸는 점 꼭 기억하세요. 첫 몇개 행을 선택하기위해서 `head()`를 사용합니다.



In [2]:
# 여기 코드를 입력하세요.
print(titanic[['PassengerId', 'Name', 'Survived']].head(10))

   PassengerId                                               Name  Survived
0            1                            Braund, Mr. Owen Harris         0
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...         1
2            3                             Heikkinen, Miss. Laina         1
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)         1
4            5                           Allen, Mr. William Henry         0
5            6                                   Moran, Mr. James         0
6            7                            McCarthy, Mr. Timothy J         0
7            8                     Palsson, Master. Gosta Leonard         0
8            9  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         1
9           10                Nasser, Mrs. Nicholas (Adele Achem)         1


2. 'Fare' 컬럼만 선택하여 첫 5행을 출력하세요.

1번 문제와 동일하지만 이 경우 대괄호를 하나만 사용해도 됩니다. 실행 결과가 이번에는 시리즈라는 것을 확인하세요.



In [3]:
print(titanic['Fare'].head(5))

0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
Name: Fare, dtype: float64


**문제 2: 행 선택**
다음 작업을 수행하는 코드를 작성하세요.
1. 10번째부터 15번째 행까지 선택하여 출력하세요.

`loc`을 사용할 수 도 있지만 여기서는 간단하게 슬라이싱을 사용해보겠습니다. 단 슬라이싱은 0부터 시작하는 위치 인덱스를 사용한다는 점을 주의해야 합니다. 위치 인덱스를 쓸 때 끝 인덱스는 제외되므로 끝 인덱스는 15로 지정했습니다.


In [4]:
print(titanic[9:15])

    PassengerId  Survived  Pclass                                  Name  \
9            10         1       2   Nasser, Mrs. Nicholas (Adele Achem)   
10           11         1       3       Sandstrom, Miss. Marguerite Rut   
11           12         1       1              Bonnell, Miss. Elizabeth   
12           13         0       3        Saundercock, Mr. William Henry   
13           14         0       3           Andersson, Mr. Anders Johan   
14           15         0       3  Vestrom, Miss. Hulda Amanda Adolfina   

       Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
9   female  14.0      1      0     237736  30.0708   NaN        C  
10  female   4.0      1      1    PP 9549  16.7000    G6        S  
11  female  58.0      0      0     113783  26.5500  C103        S  
12    male  20.0      0      0  A/5. 2151   8.0500   NaN        S  
13    male  39.0      1      5     347082  31.2750   NaN        S  
14  female  14.0      0      0     350406   7.8542   NaN        S 

2. 마지막 10개 행을 선택하여 'Name', 'Age', 'Survived' 컬럼을 출력하세요. (힌트: 음수 인덱스나 슬라이싱 활용)

이 경우 음수 인덱스를 사용하여 행에 대해 슬라이싱하고 컬럼을 선택합니다. 물론 `tail()`을 사용해도 동일한 결과를 얻을 수 있습니다.



In [5]:
print(titanic[-10:][['Name', 'Age', 'Survived']])

                                         Name   Age  Survived
881                        Markun, Mr. Johann  33.0         0
882              Dahlberg, Miss. Gerda Ulrika  22.0         0
883             Banfield, Mr. Frederick James  28.0         0
884                    Sutehall, Mr. Henry Jr  25.0         0
885      Rice, Mrs. William (Margaret Norton)  39.0         0
886                     Montvila, Rev. Juozas  27.0         0
887              Graham, Miss. Margaret Edith  19.0         1
888  Johnston, Miss. Catherine Helen "Carrie"   NaN         0
889                     Behr, Mr. Karl Howell  26.0         1
890                       Dooley, Mr. Patrick  32.0         0


**문제 3: loc과 iloc 활용**
다음 작업을 수행하는 코드를 작성하세요.
1. loc을 사용하여 인덱스 0부터 4까지 행의 'Name', 'Age', 'Fare' 컬럼을 선택하세요.

이제부터 loc, iloc을 사용합니다. 대괄호 사이에 행, 열 인덱스를 사용할 수 있고 각 인덱스 자리에 슬라이싱 표현도 사용할 수 있습니다.



In [6]:
print(titanic.loc[0:4, ['Name', 'Age', 'Fare']])

                                                Name   Age     Fare
0                            Braund, Mr. Owen Harris  22.0   7.2500
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0  71.2833
2                             Heikkinen, Miss. Laina  26.0   7.9250
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0  53.1000
4                           Allen, Mr. William Henry  35.0   8.0500


2. iloc을 사용하여 처음 3행, 처음 4열을 선택하세요.



In [7]:
print(titanic.iloc[:3, :4])

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name  
0                            Braund, Mr. Owen Harris  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                             Heikkinen, Miss. Laina  


**문제 4: 조건부 선택 (기본)**
다음 조건에 맞는 승객들을 찾는 코드를 작성하세요.
1. 3등급 승객들만 선택하여 총 몇 명인지 출력하세요.

조건 하나를 사용하여 행을 필터링합니다. 1단계 특정 열에 조건을 적용하고 그 결과로 2단계 행을 선택합니다. 항상 두 단계로 필터링한다는 점을 기억하세요.


In [8]:
# 여기 코드를 입력하세요.
print(titanic[titanic['Pclass'] == 3].shape[0])

491


2. 50세 이상 승객들을 선택하여 'Name', 'Age', 'Sex' 정보를 첫 5명만 출력하세요.


In [9]:
# 여기 코드를 입력하세요.
print(titanic[titanic['Age'] >= 50][['Name', 'Age', 'Sex']].head(5))

                                Name   Age     Sex
6            McCarthy, Mr. Timothy J  54.0    male
11          Bonnell, Miss. Elizabeth  58.0  female
15  Hewlett, Mrs. (Mary D Kingcome)   55.0  female
33             Wheadon, Mr. Edward H  66.0    male
54    Ostby, Mr. Engelhart Cornelius  65.0    male


3. 요금(Fare)이 100 이상인 승객들을 선택하여 총 몇 명인지 출력하세요.

`sum()`함수는 주어진 값을 모두 더합니다. 'Fare'에 대한 비교 결과는 True, False이므로 모두 더하면 True가 1로 취급되고 False가 0으로 취급되어 요금이 100이상인 승객의 수가 계산됩니다.


In [10]:
# 여기 코드를 입력하세요.
print(sum(titanic['Fare'] >= 100))

53


**문제 5: 조건부 선택 (복합 조건)**
다음 복합 조건에 맞는 승객들을 찾는 코드를 작성하세요.
1. 남성이면서 1등급인 승객들의 수를 구하세요.

조건이 여러개 일때 각 조건을 괄호로 반드시 감싸고 조건을 의미에 맞게 `&`, `|`로 적당히 연결하면 됩니다.



In [11]:
# 여기 코드를 입력하세요.
print(sum((titanic['Sex'] == 'male') & (titanic['Pclass'] == 1)))

122


2. 여성이면서 생존한 승객들 중 첫 3명의 'Name', 'Age', 'Pclass' 정보를 출력하세요.

아래 코드처럼 코드가 한줄에 너무 길게 진행되면 (), [] 를 기준으로 적당히 줄바꿈해서 보기 쉽게하는 것도 답안을 작성하는 요령입니다.


In [12]:
# 여기 코드를 입력하세요.
print(
    titanic[
        (titanic['Sex'] == 'female') & (titanic['Survived'] == 1)
    ].iloc[:3][['Name', 'Age', 'Pclass']]
)

                                                Name   Age  Pclass
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0       1
2                             Heikkinen, Miss. Laina  26.0       3
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0       1


3. 나이가 10세 미만이거나 60세 이상인 승객들의 수를 구하세요.



In [13]:
# 여기 코드를 입력하세요.
print(
    titanic[
        (titanic['Age'] < 10) | (titanic['Age'] >= 60)
    ].shape[0]
)

88


4. Pclass가 2등급 또는 3등급이면서 생존한 승객들 중 'Name', 'Pclass', 'Survived' 정보를 첫 5명만 출력하세요.

이 문제는 `&`와 `|`가 함께 사용되는 문제입니다. 조건이 좀 복잡하기 때문에 천천히 생각하면서 코드로 바꾸는 연습을 충분히 하도록 하세요.



In [14]:
# 여기 코드를 입력하세요.
print(
    titanic[
        ((titanic['Pclass'] == 2) | (titanic['Pclass'] == 3))
        &
        (titanic['Survived'] == 1)
    ][['Name', 'Pclass', 'Survived']].iloc[:5]
)

                                                 Name  Pclass  Survived
2                              Heikkinen, Miss. Laina       3         1
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)       3         1
9                 Nasser, Mrs. Nicholas (Adele Achem)       2         1
10                    Sandstrom, Miss. Marguerite Rut       3         1
15                   Hewlett, Mrs. (Mary D Kingcome)        2         1


**문제 6: 종합 문제**
다음 분석을 수행하세요.
1. 각 등급(Pclass)별 여성 승객의 생존율을 비교하기 위해, 1등급 여성 승객 중 생존자 수와 전체 1등급 여성 승객 수를 각각 구하여 생존률을 계산하세요.





In [15]:
# 여기 코드를 입력하세요.
titanic_female_pclass1 = titanic[(titanic['Pclass'] == 1) & (titanic['Sex'] == 'female')]
survival_rate = titanic_female_pclass1['Survived'].sum() / titanic_female_pclass1.shape[0]

print(f"{survival_rate:.2%}")

96.81%


분석 결과 1등급 여성 승객들은 대부분 생존했음을 알 수 있습니다.

2. 각 등급(Pclass)별 생존율을 계산하여 비교하세요. 1등급, 2등급, 3등급 각각의 (생존자 수 / 전체 승객 수)를 구하고, 어느 등급의 생존율이 가장 높은지 분석하세요.





In [16]:
# 여기 코드를 입력하세요.
# 1등급
pclass1 = titanic[titanic['Pclass'] == 1]
survival_rate_1 = pclass1['Survived'].sum() / pclass1.shape[0]
print(f"1등급 생존율: {survival_rate_1:.2%} ({pclass1['Survived'].sum()}/{pclass1.shape[0]})")

# 2등급
pclass2 = titanic[titanic['Pclass'] == 2]
survival_rate_2 = pclass2['Survived'].sum() / pclass2.shape[0]
print(f"2등급 생존율: {survival_rate_2:.2%} ({pclass2['Survived'].sum()}/{pclass2.shape[0]})")

# 3등급
pclass3 = titanic[titanic['Pclass'] == 3]
survival_rate_3 = pclass3['Survived'].sum() / pclass3.shape[0]
print(f"3등급 생존율: {survival_rate_3:.2%} ({pclass3['Survived'].sum()}/{pclass3.shape[0]})")

1등급 생존율: 62.96% (136/216)
2등급 생존율: 47.28% (87/184)
3등급 생존율: 24.24% (119/491)


결과를 보면 높은 등급의 객실을 이용한 승객들이 더 많이 생존했다는 사실을 명확히 확인할 수 있습니다.

3. 여성 승객과 남성 승객의 생존율을 각각 구하고, 18세 미만 어린이와 성인 남성(18세 이상 남성)의 생존율을 각각 구하세요. "여성과 어린이 먼저" 원칙이 실제로 적용되었는지 분석하세요.





In [17]:
# 여기 코드를 입력하세요.
# 여성 생존율
female = titanic[titanic['Sex'] == 'female']
female_survival = female['Survived'].sum() / female.shape[0]
print(f"여성 생존율: {female_survival:.2%} ({female['Survived'].sum()}/{female.shape[0]})")

# 남성 생존율
male = titanic[titanic['Sex'] == 'male']
male_survival = male['Survived'].sum() / male.shape[0]
print(f"남성 생존율: {male_survival:.2%} ({male['Survived'].sum()}/{male.shape[0]})")

# 어린이(18세 미만) 생존율
children = titanic[titanic['Age'] < 18]
children_survival = children['Survived'].sum() / children.shape[0]
print(f"\n어린이(18세 미만) 생존율: {children_survival:.2%} ({children['Survived'].sum()}/{children.shape[0]})")

# 성인 남성(18세 이상 남성) 생존율
adult_male = titanic[(titanic['Age'] >= 18) & (titanic['Sex'] == 'male')]
adult_male_survival = adult_male['Survived'].sum() / adult_male.shape[0]
print(f"성인 남성(18세 이상) 생존율: {adult_male_survival:.2%} ({adult_male['Survived'].sum()}/{adult_male.shape[0]})")

여성 생존율: 74.20% (233/314)
남성 생존율: 18.89% (109/577)

어린이(18세 미만) 생존율: 53.98% (61/113)
성인 남성(18세 이상) 생존율: 17.72% (70/395)


여성(74.2%)과 어린이(54.0%)의 생존율이 성인 남성(17.7%)보다 높아, '여성과 어린이 먼저' 원칙이 실제로 적용되었음을 알 수 있습니다.